### PR TITLE
[Train] Add support for handling multiple batch data types for prepare_data_loader

### LINKED ISSUE #26385: [Train] Add support for handling multiple batch data types for prepare_data_loader
### What happened + What you expected to happen

When working with Ray Train, using the `ray.train.torch.prepare_data_loader` method with a dataset that returns a dictionary instead of a tuple from its `__getitem__` method causes issues.

That's because in the `_WrappedDataLoader` class, the `_move_to_device` method only deals with a tuple:

```python
return tuple(try_move_device(i) for i in item)
```

https://github.com/ray-project/ray/blob/6dd564d8b5123be5f3cf8e95197f4866c003b312/python/ray/train/torch/train_loop_utils.py#L539-L551

Ideally, this should deal with a dictionary type batch and an arbitrary level of nesting just like Pytorch does.

### Versions / Dependencies

Python: 3.8.10
Ray: Tested with 1.13.0. Code hasn't changed in master except for a move from `torch.py -> train_loop_utils.py`

### Reproduction script

Using any dataset that returns a dictionary from its __getitem__ method should demonstrate this issue.

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

### PR DESCRIPTION
**Pull request by VishDev12**
## Why are these changes needed?

When working with Ray Train, using the `ray.train.torch.prepare_data_loader` method with a dataset that returns a dictionary instead of a tuple from its `__getitem__` method causes issues.

## Related issue number

Closes #26385

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/train/torch/train_loop_utils.py [hunk 1]
```diff
@@ -548,7 +549,18 @@ def try_move_device(i):
             return i
 
         with torch.cuda.stream(self._memcpy_stream):
-            return tuple(try_move_device(i) for i in item)
+            if isinstance(item, collections.abc.Mapping):
+                item_on_device = {k: self._move_to_device(v) for k, v in item.items()}
+            elif isinstance(item, (tuple, list)):
+                item_on_device = tuple(try_move_device(i) for i in item)
```
[the remark this answers is not in the document]
Comment (VishDev12) [reply python/ray/train/torch/train_loop_utils.py:555]: Great point, I hadn't considered that, I'll commit this

File: python/ray/train/torch/train_loop_utils.py [hunk 2]
```diff
@@ -548,7 +549,18 @@ def try_move_device(i):
...
         with torch.cuda.stream(self._memcpy_stream):
-            return tuple(try_move_device(i) for i in item)
+            if isinstance(item, collections.abc.Mapping):
+                item_on_device = {k: self._move_to_device(v) for k, v in item.items()}
+            elif isinstance(item, (tuple, list)):
+                item_on_device = tuple(self._move_to_device(i) for i in item)
+            elif isinstance(item, torch.Tensor):
+                item_on_device = try_move_device(item)
+            else:
+                logger.debug(
```
Comment (amogkam) [inline python/ray/train/torch/train_loop_utils.py:559]: Should we use info here instead?

**Review by amogkam — COMMENTED** [review]
Thanks @VishDev12! Do you think you can also add a test for this? There's already a `test_torch_prepare_dataloader` test in `test_gpu.py`. We can just use that same test but try with DataLoaders that also output dictionaries instead of just tuples.

**Comment by VishDev12** [comment]
Hey @amogkam, I've added a dict case in the `test_torch_prepare_dataloader` test.

Please let me know if something doesn't look quite right.

File: python/ray/train/examples/torch_linear_example.py [hunk 3]
```diff
@@ -24,6 +24,11 @@ def __len__(self):
         return len(self.x)
 
 
+class LinearDatasetDict(LinearDataset):
```
Comment (amogkam) [inline python/ray/train/examples/torch_linear_example.py:27]: If this is only used in the test and not part of the example, can we move this to `test_gpu` directly?

**Review by amogkam — APPROVED** [review]
Thanks @VishDev12 this lgtm! Just left one minor comment

File: python/ray/train/examples/torch_linear_example.py (hunk 3 above)
Comment (VishDev12) [reply python/ray/train/examples/torch_linear_example.py]: That makes sense, I've moved it just under the fixtures defined on `test_gpu`, just to keep the separation between the `test_*` functions and the rest. But if you want to have it defined just above `test_torch_prepare_dataloader`, I could do that too.

**Comment by amogkam** [comment]
Awesome, thanks @VishDev12!

**Pull request merged**